File: sdk/cpp/test/codec_test.cpp
#include "cap/codec.h"

#include <string>
#include <vector>

#include <gtest/gtest.h>

#include "cordum/agent/v1/buspacket.pb.h"
#include "cordum/agent/v1/job.pb.h"
#include "cordum/agent/v1/heartbeat.pb.h"

namespace {

cordum::agent::v1::BusPacket MakePacket() {
  cordum::agent::v1::BusPacket packet;
  packet.set_trace_id("test-trace");
  packet.set_sender_id("test-sender");
  packet.set_protocol_version(1);
  auto* req = packet.mutable_job_request();
  req->set_job_id("job-1");
  req->set_topic("test.topic");
  return packet;
}

}  // namespace

TEST(CodecTest, DeterministicSerializationIsStable) {
  auto packet = MakePacket();

  auto first = cap::MarshalDeterministic(packet);
  ASSERT_FALSE(first.empty());

  for (int i = 0; i < 100; ++i) {
    auto next = cap::MarshalDeterministic(packet);
    ASSERT_EQ(first.size(), next.size()) << "iteration " << i;
    EXPECT_EQ(first, next) << "iteration " << i;
  }
}

TEST(CodecTest, DeterministicHeartbeatIsStable) {
  cordum::agent::v1::BusPacket packet;
  packet.set_sender_id("worker-1");
  packet.set_protocol_version(1);
  auto* hb = packet.mutable_heartbeat();
  hb->set_worker_id("worker-1");
  hb->set_pool("pool-a");
  hb->set_active_jobs(3);
  hb->set_max_parallel_jobs(8);
  hb->set_cpu_load(42.5);
  hb->set_memory_load(55.0);
  hb->set_progress_pct(67);
  hb->set_last_memo("checkpoint");

  auto first = cap::MarshalDeterministic(packet);
  ASSERT_FALSE(first.empty());

  for (int i = 0; i < 50; ++i) {
    EXPECT_EQ(first, cap::MarshalDeterministic(packet));
  }
}

TEST(CodecTest, MarshalUnsignedForSignatureStripsSignature) {
  auto packet = MakePacket();
  packet.set_signature("fake-sig-data");

  auto bytes = cap::MarshalUnsignedForSignature(packet);
  ASSERT_FALSE(bytes.empty());

  // Deserialize and verify signature is gone
  cordum::agent::v1::BusPacket decoded;
  ASSERT_TRUE(decoded.ParseFromArray(bytes.data(), static_cast<int>(bytes.size())));
  EXPECT_TRUE(decoded.signature().empty());
  EXPECT_EQ(decoded.sender_id(), "test-sender");
  EXPECT_EQ(decoded.trace_id(), "test-trace");
  EXPECT_EQ(decoded.protocol_version(), 1);
  EXPECT_EQ(decoded.job_request().job_id(), "job-1");
}

TEST(CodecTest, MarshalUnsignedPreservesAllOtherFields) {
  auto packet = MakePacket();
  packet.set_signature("to-be-stripped");

  auto without_sig = cap::MarshalUnsignedForSignature(packet);

  // Also serialize a copy with no signature
  cordum::agent::v1::BusPacket clean;
  clean.CopyFrom(packet);
  clean.clear_signature();
  auto clean_bytes = cap::MarshalDeterministic(clean);

  EXPECT_EQ(without_sig, clean_bytes);
}

TEST(CodecTest, EmptyMessage) {
  cordum::agent::v1::BusPacket empty;
  auto bytes = cap::MarshalDeterministic(empty);
  // Empty protobuf message serializes to zero bytes
  EXPECT_TRUE(bytes.empty());
}

TEST(CodecTest, RoundTrip) {
  auto packet = MakePacket();
  auto bytes = cap::MarshalDeterministic(packet);
  ASSERT_FALSE(bytes.empty());

  cordum::agent::v1::BusPacket decoded;
  ASSERT_TRUE(decoded.ParseFromArray(bytes.data(), static_cast<int>(bytes.size())));
  EXPECT_EQ(decoded.trace_id(), packet.trace_id());
  EXPECT_EQ(decoded.sender_id(), packet.sender_id());
  EXPECT_EQ(decoded.protocol_version(), packet.protocol_version());
  EXPECT_EQ(decoded.job_request().job_id(), packet.job_request().job_id());
  EXPECT_EQ(decoded.job_request().topic(), packet.job_request().topic());
}

File: sdk/java/src/test/java/io/cordum/cap/CodecTest.java
package io.cordum.cap;

import com.google.protobuf.ByteString;
import com.google.protobuf.Timestamp;
import io.cordum.cap.agent.v1.BusPacket;
import io.cordum.cap.agent.v1.JobRequest;
import org.junit.jupiter.api.Test;

import static org.junit.jupiter.api.Assertions.*;

class CodecTest {

    private BusPacket makePacket() {
        return BusPacket.newBuilder()
                .setTraceId("test-trace")
                .setSenderId("test-sender")
                .setProtocolVersion(1)
                .setJobRequest(JobRequest.newBuilder()
                        .setJobId("job-1")
                        .setTopic("test.topic")
                        .build())
                .build();
    }

    @Test
    void deterministicSerializationIsStable() {
        BusPacket packet = makePacket();
        byte[] first = Codec.marshalDeterministic(packet);
        assertNotNull(first);
        assertTrue(first.length > 0);

        for (int i = 0; i < 100; i++) {
            byte[] next = Codec.marshalDeterministic(packet);
            assertArrayEquals(first, next, "iteration " + i);
        }
    }

    @Test
    void deterministicHeartbeatIsStable() {
        io.cordum.cap.agent.v1.Heartbeat hb = io.cordum.cap.agent.v1.Heartbeat.newBuilder()
                .setWorkerId("worker-1")
                .setPool("pool-a")
                .setActiveJobs(3)
                .setMaxParallelJobs(8)
                .setCpuLoad(42.5f)
                .setMemoryLoad(55.0f)
                .setProgressPct(67)
                .setLastMemo("checkpoint")
                .build();

        BusPacket packet = BusPacket.newBuilder()
                .setSenderId("worker-1")
                .setProtocolVersion(1)
                .setHeartbeat(hb)
                .build();

        byte[] first = Codec.marshalDeterministic(packet);
        for (int i = 0; i < 50; i++) {
            assertArrayEquals(first, Codec.marshalDeterministic(packet));
        }
    }

    @Test
    void marshalUnsignedStripsSignature() {
        BusPacket packet = makePacket().toBuilder()
                .setSignature(ByteString.copyFromUtf8("fake-sig-data"))
                .build();

        byte[] bytes = Codec.marshalUnsignedForSignature(packet);
        assertTrue(bytes.length > 0);

        BusPacket decoded = assertDoesNotThrow(() -> BusPacket.parseFrom(bytes));
        assertTrue(decoded.getSignature().isEmpty());
        assertEquals("test-sender", decoded.getSenderId());
        assertEquals("test-trace", decoded.getTraceId());
        assertEquals(1, decoded.getProtocolVersion());
        assertEquals("job-1", decoded.getJobRequest().getJobId());
    }

    @Test
    void marshalUnsignedPreservesAllOtherFields() {
        BusPacket packet = makePacket().toBuilder()
                .setSignature(ByteString.copyFromUtf8("to-be-stripped"))
                .build();

        byte[] withoutSig = Codec.marshalUnsignedForSignature(packet);
        byte[] clean = Codec.marshalDeterministic(
                packet.toBuilder().clearSignature().build());

        assertArrayEquals(clean, withoutSig);
    }

    @Test
    void emptyMessage() {
        BusPacket empty = BusPacket.getDefaultInstance();
        byte[] bytes = Codec.marshalDeterministic(empty);
        assertEquals(0, bytes.length);
    }

    @Test
    void roundTrip() {
        BusPacket packet = makePacket();
        byte[] bytes = Codec.marshalDeterministic(packet);
        assertTrue(bytes.length > 0);

        BusPacket decoded = assertDoesNotThrow(() -> BusPacket.parseFrom(bytes));
        assertEquals(packet.getTraceId(), decoded.getTraceId());
        assertEquals(packet.getSenderId(), decoded.getSenderId());
        assertEquals(packet.getProtocolVersion(), decoded.getProtocolVersion());
        assertEquals(packet.getJobRequest().getJobId(), decoded.getJobRequest().getJobId());
        assertEquals(packet.getJobRequest().getTopic(), decoded.getJobRequest().getTopic());
    }
}
